### DISCUSSION TITLE
SqlAlchemy database adapter code base

### DISCUSSION THREAD

**Discussion by Tobi-De** [fastapi-users/fastapi-users#995] [Q&A] [answered]
Hi @frankie567, I hope you are well. I have a quick question for you.

```python
@declarative_mixin
class SQLAlchemyBaseUserTable(Generic[ID]):
    """Base SQLAlchemy users table definition."""

    __tablename__ = "user"

    if TYPE_CHECKING:  # pragma: no cover
        id: ID
        email: str
        hashed_password: str
        is_active: bool
        is_superuser: bool
        is_verified: bool
    else:
        email: str = Column(String(length=320), unique=True, index=True, nullable=False)
        hashed_password: str = Column(String(length=1024), nullable=False)
        is_active: bool = Column(Boolean, default=True, nullable=False)
        is_superuser: bool = Column(Boolean, default=False, nullable=False)
        is_verified: bool = Column(Boolean, default=False, nullable=False)

```

I was wondering why the  `if TYPE_CHEKING`  code block? The types are already defined in the else clause, there is probably something I don't understand about type checkers or maybe it's related to sqlachemy itself, can you clarify that for me?

**Comment by frankie567** [comment, marked as answer]
Hi @Tobi-De 👋 

So, in theory, you're 100% correct, since the `Column` attributes are typed, it should work. However, there are lot of quirks with typing regarding SQLAlchemy.

But let's get back to the beginning: FastAPI Users defines a `Protocol` that a User class model should conform to:

https://github.com/fastapi-users/fastapi-users/blob/2477fdffc118f9d75b9031f236037e57e058dbaa/fastapi_users/models.py#L12-L23

Basically, we tell the type checker to verify that we give to `BaseUserDatabase` a class that has those properties. What's nice with `Protocol` is that we don't have to explicitly inherit from it, the type checker will enforce the constraint by itself.

Now, let's implement this model with SQLAlchemy. The most natural way would be this, as you have guessed:

```py
class SQLAlchemyBaseUserTable(Generic[ID]):
    """Base SQLAlchemy users table definition."""

    __tablename__ = "user"

    email: str = Column(String(length=320), unique=True, index=True, nullable=False)
    hashed_password: str = Column(String(length=1024), nullable=False)
    is_active: bool = Column(Boolean, default=True, nullable=False)
    is_superuser: bool = Column(Boolean, default=False, nullable=False)
    is_verified: bool = Column(Boolean, default=False, nullable=False)
```

However, when doing this, we get an error from the type checker:

```
fastapi_users_db_sqlalchemy/__init__.py:34: error: Incompatible types in assignment (expression has type "Column[String]", variable has type "str")
fastapi_users_db_sqlalchemy/__init__.py:35: error: Incompatible types in assignment (expression has type "Column[String]", variable has type "str")
fastapi_users_db_sqlalchemy/__init__.py:36: error: Incompatible types in assignment (expression has type "Column[Boolean]", variable has type "bool")
fastapi_users_db_sqlalchemy/__init__.py:37: error: Incompatible types in assignment (expression has type "Column[Boolean]", variable has type "bool")
fastapi_users_db_sqlalchemy/__init__.py:38: error: Incompatible types in assignment (expression has type "Column[Boolean]", variable has type "bool")
```

Logically, mypy doesn't understand why we tell it that a `Column` instance is a `str` or a `bool`. SQLAlchemy construct is indeed quite special here.

Fortunately, SQLAlchemy provides a [special mypy plugin](https://docs.sqlalchemy.org/en/14/orm/extensions/mypy.html) to help it understand all its specificities. In particular, it provides the [decorator `declarative_mixin`](https://docs.sqlalchemy.org/en/14/orm/extensions/mypy.html#using-declared-attr-and-declarative-mixins), which is useful when defining base classes that will be used by actual models. So now we can write this:

```py
@declarative_mixin
class SQLAlchemyBaseUserTable(Generic[ID]):
    """Base SQLAlchemy users table definition."""

    __tablename__ = "user"

    email: str = Column(String(length=320), unique=True, index=True, nullable=False)
    hashed_password: str = Column(String(length=1024), nullable=False)
    is_active: bool = Column(Boolean, default=True, nullable=False)
    is_superuser: bool = Column(Boolean, default=False, nullable=False)
    is_verified: bool = Column(Boolean, default=False, nullable=False)
```

This solves the previous error, but now mypy complains because it's not compliant with `UserProtocol`:

```
tests/test_users.py:72: error: Type argument "User" of "SQLAlchemyUserDatabase" must be a subtype of "UserProtocol[Any]"
```

Now, I don't really know *why* it happens. My theory is that the SQLAlchemy mypy plugin is doing weird things with the class that prevent mypy to infer the protocol correctly (maybe it still tells it the attributes are `Column`).

So... The only solution I found is the one you showed in your post: a special version for `TYPE_CHECKING` to force the type of each attributes.

Maybe there are better solutions, but typing with SQLAlchemy is quite complicated. Apparently, they will get rid of the mypy plugin in their next version, so maybe it'll solve this kind of issue.

Hope it clarifies things 😃  Cheers!

**Reply by Tobi-De** [reply]
Thank you very much for the detailed answer, I'm trying to please mypy on my tortoise-orm adapter, it's a nightmare. 😆

**Reply by frankie567** [reply]
Yeah, I suppose it's same kind of issues 😅

**Comment by Tobi-De** [comment]
might as well take the opportunity to ask, @frankie567  do you think it make sense  to use the protocols (UP, AP, etc....) here ?
```py
class TortoiseBaseAccessTokenModel(models.Model):
    """Tortoise access token model definition."""

    token = fields.CharField(pk=True, max_length=43, null=False)
    created_at: datetime = fields.DatetimeField(null=False, auto_now_add=True)
    user = fields.ForeignKeyField("models.User")

    class Meta:
        abstract = True


class TortoiseAccessTokenDatabase(Generic[AP], AccessTokenDatabase[AP]):
    """
    Access token database adapter for Tortoise ORM.
    :param access_token_model: Access token Tortoise ORM model.
    """

    def __init__(self, access_token_model: Type[AP]):
        self.access_token_model = access_token_model
```

The methods on the class `TortoiseAccessTokenDatabase` depends on the `access_token_model` being a tortoise orm model and having methods like `filter`, `get`etc..., I think in this case it doesn't really make sens to use generics, the `access_token_model` has to inherit from the `TortoiseBaseAccessTokenModel` anyway, so I think it is probably better to use that as the expected `Type`, what do you think ?

**Reply by frankie567** [reply]
Yes, it could work, I think. That said might worth to check what I did for Beanie: https://github.com/fastapi-users/fastapi-users-db-beanie/blob/main/fastapi_users_db_beanie/__init__.py

Basically, I defined a "sub-generic" type bound to a base model from ORM, from which we can have all the ORM methods. 

I started to do the same for SQAlchemy but I can't remember why I gave up. I might have another try 😅

**Reply by Tobi-De** [reply]
Thanks, I thought of something like that too, the only reason I haven't tried it is pure laziness, I'll get to it.
